test/controlplane/policies/k8s/policy.go

```diff
@@ -0,0 +1,129 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package k8s
+
+import (
+	"fmt"
+	"os"
+	"path"
+	"testing"
+
+	k8sConst "github.com/cilium/cilium/pkg/k8s/apis/cilium.io"
+	"github.com/cilium/cilium/pkg/labels"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/cilium/cilium/pkg/policy"
+	"github.com/cilium/cilium/pkg/policy/api"
+	"github.com/cilium/cilium/test/controlplane/suite"
+)
+
+func init() {
+	suite.AddTestCase("Policies/k8s", testK8SNetworkPolicy)
+}
+
+type direction int
+
+const (
+	egress = direction(iota)
+	ingress
+	both
+)
+
+type step struct {
+	filename         string
+	namespace        string
+	direction        direction
+	expectedDecision api.Decision
+}
+
+var steps = []step{
+	{
+		filename:         "state1.yaml",
+		namespace:        "policy1",
+		direction:        egress,
+		expectedDecision: api.Allowed,
+	},
+	{
+		filename:         "state2.yaml",
+		namespace:        "policy2",
+		direction:        ingress,
+		expectedDecision: api.Allowed,
+	},
+	{
+		filename:         "state3.yaml",
+		namespace:        "policy3",
+		direction:        egress,
+		expectedDecision: api.Denied,
+	},
+	{
+		filename:         "state4.yaml",
+		namespace:        "policy4",
+		direction:        ingress,
+		expectedDecision: api.Denied,
+	},
+	{
+		filename:         "state5.yaml",
+		namespace:        "policy5",
+		direction:        both,
+		expectedDecision: api.Denied,
+	},
+}
+
+func testK8SNetworkPolicy(t *testing.T) {
+	cwd, err := os.Getwd()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	modConfig := func(c *option.DaemonConfig) {
+		c.Debug = true
+	}
+
+	for _, version := range []string{"1.20", "1.22", "1.24"} {
+		abs := func(f string) string { return path.Join(cwd, "policies", "k8s", "v"+version, f) }
+
+		t.Run("v"+version, func(t *testing.T) {
+			test := suite.NewControlPlaneTest(t, "policy-control-plane", version)
+
+			// Feed in initial state and start the agent.
+			test.UpdateObjectsFromFile(abs("init.yaml")).StartAgent(modConfig)
+
+			// Run through the test steps
+			for _, step := range steps {
+				test.UpdateObjectsFromFile(abs(step.filename))
+				test.Eventually(func() error { return validate(test, step) })
+			}
+
+			test.StopAgent()
+		})
+	}
+}
+
+func validate(test *suite.ControlPlaneTest, step step) error {
+	sc := &policy.SearchContext{
+		From: labels.LabelArray{
+			labels.NewLabel(k8sConst.PodNamespaceLabel, step.namespace, labels.LabelSourceK8s),
+		},
+		To: labels.LabelArray{
+			labels.NewLabel(k8sConst.PodNamespaceLabel, step.namespace, labels.LabelSourceK8s),
+		},
+	}
+
+	repo := test.GetAgent().GetPolicyRepository()
+
+	repo.Mutex.RLock()
+	ingressDecision := test.GetAgent().GetPolicyRepository().AllowsIngressRLocked(sc)
+	egressDecision := test.GetAgent().GetPolicyRepository().AllowsEgressRLocked(sc)
+	repo.Mutex.RUnlock()
```

Comment:
I think the key question is how far we hope to take this test framework. Right now the above is setting from/to to be the same namespace, so as a "policy test for traffic to/from the same namespace" it seems like it accurately captures the allow/deny behaviour.

There's three dimensions that make this more complicated:
1. Narrowing down the traffic types to specific ports / labels
1. Narrowing down the policy types to specific ports / labels / L7 etc.
1. Merging overlapping statements from different policies.

Broadly I think that for these more specific dimensions, we have fairly good test coverage in `pkg/policy/distillery_test.go` (based on [this test plan](https://docs.google.com/spreadsheets/d/1WANIoZGB48nryylQjjOw6lKjI80eVgPShrdMTMalLEw/edit?usp=sharing)). These tests define the policy rules directly in the test code by creating CNP rules with the API library and then testing specific combinations to evaluate the kinds of BPF map keys / values that are generated, so they don't test the parsing of rules from KNP / CNP / CCNP into the internal API format, but they test the way that rules can be combined.